---
title: "Next.js で Apollo Client のエラーをハンドリングする"
emoji: "🪐"
type: "tech" # tech: 技術記事 / idea: アイデア
topics: ["graphql"]
published: false
---

```tsx
const [addBook,{error}] = useMutation(MUTATION)

...

useEffect(()=>{
    addBook({variables:{input:{id,name,price}}})
},[])

useEffect(()=>{
    if(!error)return
    useErrorSnackbar(error.message) // エラー用のSnackbarを表示してユーザーにエラーを知らせる
},[error])
```

上記は本の情報を更新するための GraphQL mutation を実行する関数を作成し、実際にそれをレンダリング時に呼び出す実装をしています。
このとき、既に登録済みのものの場合エラーになり、Snackbar で「既に登録されています」というエラーが表示される挙動が期待されるのですが、実際には `Unhandled Runtime Error`が発生し、ブラウザ上で Next.js のエラーを表示するモーダルがオーバーレイして表示されます。
そこで Apollo Client のエラーハンドリングの仕組みについて調べてみます。

## エラーの種類

`GraphQL Error`と`network error`の二種類があります。

### GraphQL Error

- Syntax error
  - クエリやミューテーションの文法記法に関する場合などに発生
- Validation error
  - クエリやミューテーションに存在しないフィールドがある場合などに発生
- Resolver error
  - クエリに値を入力する際などに発生したエラー

構文エラーやバリデーションエラー発生時はサーバー側でクエリ等は実行されません。GraphQL Error が発生した場合、サーバーはそのエラーを Apollo Client へのレスポンスの errors 配列に含めます。
リゾルバエラーが発生し、レスポンスに部分的なデータが含まれている場合、場合、HTTP ステータスコードは`200 OK`で応答します。この時の部分的なデータとは、操作で要求したデータの一部などのことを指します。クエリの結果が 1 つまたは複数のエラーになった場合、`undefined`になることがあります。
Apollo Client はデフォルトでこの部分データを無視しますが、GraphQL エラーポリシーを設定することでこの動作をオーバーライドすることができます。

### エラーポリシー

- `none`
  - レスポンスに GraphQL エラーが含まれる場合、`error.graphQLErrors`で返され、サーバーがレスポンスでデータを返してもレスポンスデータは undefined に設定されます。ネットワークエラーと GraphQL エラーは、デフォルトでは同様のレスポンスの形式になります。
- `ignore`
  - graphQLErrors は無視され(`error.graphQLErrors`には何も返されません)、返されたデータはキャッシュされ、あたかもエラーが発生しなかったかのようにレンダリングされます。
- `all`
  - data と error.graphQLErrors の両方が入力され、結果の一部とエラー情報の両方をレンダリングすることが可能です。
